diff --git a/xbmc/pvr/PVRManager.cpp b/xbmc/pvr/PVRManager.cpp
index f1bca75..3504920 100644
--- a/xbmc/pvr/PVRManager.cpp
+++ b/xbmc/pvr/PVRManager.cpp
@@ -368,7 +368,9 @@ void CPVRManager::Cleanup(void)
 {
   CSingleLock lock(m_critSection);
 
-  SAFE_DELETE(m_addons);
+  CPVRClients *addons = m_addons;
+  m_addons = nullptr;
+
   SAFE_DELETE(m_guiInfo);
   SAFE_DELETE(m_timers);
   SAFE_DELETE(m_recordings);
@@ -395,6 +397,11 @@ void CPVRManager::Cleanup(void)
   HideProgressDialog();
 
   SetState(ManagerStateStopped);
+
+  // Destructing addons with locked PVR manager mutex may lead to deadlocks due
+  // to callbacks from the addons to PVR manager.
+  lock.Leave();
+  SAFE_DELETE(addons);
 }
 
 void CPVRManager::ResetProperties(void)
